File: python/src/liberty/_realtime.py
"""Sessão de transcrição ao vivo.

A rota é WebSocket, não HTTP, então ela não cabe em `audio.transcriptions`. O
que muda para quem integra é a latência: o arquivo só volta texto no fim, e a
sessão devolve trecho enquanto a pessoa fala.

O protocolo tem quatro passos: abrir, configurar a sessão, mandar áudio em
pedaços e ler os eventos de volta. `transcribe_stream` faz os quatro.
"""

from __future__ import annotations

import asyncio
import base64
from typing import TYPE_CHECKING, Any, AsyncIterable, AsyncIterator, cast

from ._errors import TranscriptionFailed

if TYPE_CHECKING:
    from ._client import AsyncLiberty

DELTA = "conversation.item.input_audio_transcription.delta"
CONCLUIDO = "conversation.item.input_audio_transcription.completed"
FALHOU = "conversation.item.input_audio_transcription.failed"


async def transcribe_stream(
    client: AsyncLiberty,
    audio: AsyncIterable[bytes],
    *,
    model: str,
    language: str | None = None,
    prompt: str | None = None,
    sample_rate: int = 24000,
) -> AsyncIterator[str]:
    """Transcreve áudio ao vivo e devolve o texto em pedaços.

    `audio` entrega PCM16 mono cru. Cada pedaço que sai dele é um trecho
    fechado: o cliente manda e pede a transcrição daquele trecho, então quem
    decide onde cortar é quem produz o áudio. Corte na pausa da fala.

        async for texto in client.transcribe_stream(pedacos, model="liberty-<apelido-de-transcricao>"):
            print(texto, end="", flush=True)

    Levanta `TranscriptionFailed` se a sessão recusar o áudio.
    """
    transcricao: dict[str, Any] = {"model": model}
    if language:
        transcricao["language"] = language
    if prompt:
        transcricao["prompt"] = prompt

    # `intent=transcription` no lugar de `model=`: esta rota recusa modelo na
    # query, e quem resolve o modelo é o apelido dentro da sessão.
    async with client.realtime.connect(extra_query={"intent": "transcription"}) as conexao:
        await conexao.session.update(
            session=cast(
                Any,
                {
                    "type": "transcription",
                    "audio": {
                        "input": {
                            "format": {"type": "audio/pcm", "rate": sample_rate},
                            "transcription": transcricao,
                            # server_vad não vale nesta rota: quem corta o
                            # trecho é quem manda o áudio.
                            "turn_detection": None,
                        }
                    },
                },
            )
        )

        enviados = 0
        fim_do_audio = asyncio.Event()

        async def enviar() -> None:
            nonlocal enviados
            try:
                async for pedaco in audio:
                    if not pedaco:
                        continue
                    await conexao.input_audio_buffer.append(
                        audio=base64.b64encode(pedaco).decode()
                    )
                    await conexao.input_audio_buffer.commit()
                    enviados += 1
            finally:
                fim_do_audio.set()
                # Sem trecho enviado não vem evento nenhum, e o laço de leitura
                # esperaria para sempre. Fechar encerra a iteração.
                if enviados == 0:
                    await conexao.close()

        produtor = asyncio.create_task(enviar())
        concluidos = 0
        try:
            async for evento in conexao:
                if evento.type == DELTA:
                    yield evento.delta
                elif evento.type == CONCLUIDO:
                    concluidos += 1
                    if fim_do_audio.is_set() and concluidos >= enviados:
                        break
                elif evento.type == FALHOU:
                    raise TranscriptionFailed(str(getattr(evento, "error", evento)))
                elif evento.type == "error":
                    raise TranscriptionFailed(str(getattr(evento, "error", evento)))
            await produtor
        finally:
            produtor.cancel()

File: node/src/realtime.ts
/**
 * Sessão de transcrição ao vivo.
 *
 * A rota é WebSocket, não HTTP, então ela não cabe em `audio.transcriptions`.
 * O que muda para quem integra é a latência: o arquivo só volta texto no fim, e
 * a sessão devolve trecho enquanto a pessoa fala.
 *
 * O helper de realtime do `openai` não serve aqui: ele exige `model` e o crava
 * na query, e esta rota recusa modelo na query. Por isso o socket é montado
 * neste módulo, com `realtimeURL()` e `realtimeHeaders()` do cliente.
 */

import WebSocket from 'ws';

import type { Liberty } from './client.js';
import { TranscriptionFailed } from './errors.js';

const DELTA = 'conversation.item.input_audio_transcription.delta';
const CONCLUIDO = 'conversation.item.input_audio_transcription.completed';
const FALHOU = 'conversation.item.input_audio_transcription.failed';

export type TranscribeOptions = {
  /** Apelido do gateway, por exemplo `liberty-<apelido-de-transcricao>`. */
  model: string;
  language?: string;
  prompt?: string;
  /** Taxa do PCM16 mono que você manda. */
  sampleRate?: number;
};

export async function* transcribeStream(
  client: Liberty,
  audio: AsyncIterable<Uint8Array>,
  { model, language, prompt, sampleRate = 24000 }: TranscribeOptions,
): AsyncGenerator<string> {
  const transcription: Record<string, string> = { model };
  if (language) transcription['language'] = language;
  if (prompt) transcription['prompt'] = prompt;

  const ws = new WebSocket(client.realtimeURL(), { headers: client.realtimeHeaders() });

  const fila: string[] = [];
  let acordar: (() => void) | null = null;
  let falha: Error | null = null;
  let terminou = false;
  let enviados = 0;
  let concluidos = 0;
  let audioAcabou = false;

  const sinalizar = () => {
    const pendente = acordar;
    acordar = null;
    pendente?.();
  };

  const encerrarSePossivel = () => {
    if (audioAcabou && concluidos >= enviados) {
      terminou = true;
      sinalizar();
    }
  };

  ws.on('message', (bruto) => {
    let evento: { type?: string; delta?: string; error?: unknown };
    try {
      evento = JSON.parse(bruto.toString());
    } catch {
      return;
    }
    if (evento.type === DELTA && typeof evento.delta === 'string') {
      fila.push(evento.delta);
      sinalizar();
    } else if (evento.type === CONCLUIDO) {
      concluidos += 1;
      encerrarSePossivel();
    } else if (evento.type === FALHOU || evento.type === 'error') {
      falha = new TranscriptionFailed(JSON.stringify(evento.error ?? evento));
      sinalizar();
    }
  });
  ws.on('error', (erro) => {
    falha = erro;
    sinalizar();
  });
  ws.on('close', () => {
    terminou = true;
    sinalizar();
  });

  await new Promise<void>((resolver, rejeitar) => {
    ws.once('open', resolver);
    ws.once('error', rejeitar);
  });

  ws.send(
    JSON.stringify({
      type: 'session.update',
      session: {
        type: 'transcription',
        audio: {
          input: {
            format: { type: 'audio/pcm', rate: sampleRate },
            transcription,
            // server_vad cortaria o trecho no servidor, e aqui quem corta é
            // quem manda o áudio.
            turn_detection: null,
          },
        },
      },
    }),
  );

  const produtor = (async () => {
    for await (const pedaco of audio) {
      if (!pedaco.length) continue;
      ws.send(
        JSON.stringify({
          type: 'input_audio_buffer.append',
          audio: Buffer.from(pedaco).toString('base64'),
        }),
      );
      ws.send(JSON.stringify({ type: 'input_audio_buffer.commit' }));
      enviados += 1;
    }
    audioAcabou = true;
    // Sem trecho não vem evento nenhum, e o laço de leitura esperaria para
    // sempre. Fechar encerra a iteração.
    if (enviados === 0) ws.close();
    encerrarSePossivel();
  })();

  try {
    for (;;) {
      while (fila.length) yield fila.shift() as string;
      if (falha) throw falha;
      if (terminou) break;
      await new Promise<void>((resolver) => {
        acordar = resolver;
      });
    }
    await produtor;
  } finally {
    ws.close();
  }
}
